### ISSUE TITLE
RfC:Switch to `cython_<lapack,blas>` based wrappers and deprecate custom `scipy.linalg.<lapack,blas>`

### ISSUE THREAD

**Issue by ilayn** [scipy/scipy#20682]
Historically, we have gone through many stages of BLAS/LAPACK wrappers. And consequently we have a lot of home-baked wrappers to the formal BLAS/LAPACK routines. 

This actually allowed the users to conveniently call native routines through `f2py` glue. While this is very valuable, our way of carrying the historical inconsistencies have been quite high. Just to name a few

- Character switches sometimes kept ala `uplo='L'` or converted into `0`, `1` ala `lower=1` (since bool is not really available). 
- The pivot arrays sometimes converted at the wrapper with `C` injection, sometimes kept in Fortran order.
- `lwork` queries sometimes included as separate public function `<routine>_lwork` but sometimes left to user by calling the same routine twice with exposed `lwork` keyword set to `-1`.
- This would also help us handling the OpenBLAS linkage or other libs a bit more ergonomic.

I think we can switch to cython based linkage to have more automated wrapper generation and less strain on our testing framework. And also it would make things more straightforward to compile. 

On the downside, we let f2py to handle the normalization of arguments and this has to be robust in the cythonized wrappers. Moreover, we will have to make sure that the F-contiguous and `<s,d,c,z>` flavorizing happens properly. 

We touched upon this many times over the years but let's consolidate the discussions here (last time we mentioned this was I think https://github.com/scipy/scipy/pull/13663#issuecomment-1179792099) 

One particular difficulty is that if we want to solve the historical issues, unfortunately, we have to get rid of all wrappers and offer new ones where the naming would be different. I'm not necessarily a big fan of this but doing this without breaking anybody's current code this seems like the only option. And probably needs more than 2 versions for deprecation cycles. 

Hence it would be the case that `scipy.linalg.lapack` and `scipy.linalg.blas` gets deprecated goes away and new ones, whatever the name might be will be the new auto-generated submodule names.

**Comment by ev-br** [comment]
cross-ref https://github.com/scipy/scipy/issues/4516

**Comment by ilayn** [comment]
Just occured to me that we can at least do this for our own internal use and leave the deprecation for later.

**Comment by rgommers** [comment]
This proposal isn't very clear to me I'm afraid. We should consider public API we expose separately from what we do internally. For public API:
- `scipy.linalg.blas` and `scipy.linalg.lapack` are quite low-level indeed. If someone wants to write nicer wrappers, then sure why not.
    - Once those exist, we can consider if the old ones can be deprecated.
- `scipy.linalg.cython_blas`/`cython_lapack` are LP64 only, and that cannot be changed without breaking backwards compatibility. We could add a second set for ILP64 builds perhaps, but that isn't done today and I haven't seen anyone planning to work on this.

For SciPy's internal usage:
- We're using the `f2py`-based wrappers internally Since we also want ILP64 support for SciPy itself, the `f2py`-based wrappers must be kept.
- The machinery to generate the wrappers was improved a fair bit recently.

> Just occured to me that we can at least do this for our own internal use

I don't think so, because of what I wrote above.

**Comment by rgommers** [comment]
> How does Cython set pickup ILP64 if the linked OpenBLAS is 32bit?

It doesn't, that was my point. Internally we can use ILP64 everywhere when the user builds for that, but the requirements for the Cython API are different.

> having two copies of OpenBLAS

Yeah, we probably don't want to put that back. That was a really weird construction. We should have a unified build that provides both LP64 and ILP64 symbols. Currently MKL and Accelerate provide that. OpenBLAS could in the future, but not today.

**Comment by ilayn** [comment]
Ah OK, intel-lingo for 32 bit is LP64, I didn't pay attention. We need to do something about this at some point anyways. I'll try some about it at when I can spare time.

**Comment by chillenb** [comment]
If you deprecate the F2PY wrappers, would it be possible to expose the Cython ones as Python functions? [Something along these lines.](https://github.com/chillenb/pycblas/blob/73562413b1bf37a1adfb8e695099c8c38b3c4411/src/pycblas.pyx). The ability to call BLAS from Python is really nice, especially `?trsm`, which is faster than scipy.linalg.solve_triangular (`?trtrs`) when B is row-major and nrhs is large. Also `?axpy`, which is much faster than single-threaded addition of ndarrays. If these wrappers are Cython-only, then they cannot be used interactively and packages that use them will need to add a Cython dependency.

**Comment by rgommers** [comment]
> If you deprecate the F2PY wrappers, 

I don't think we will. At least I haven't seen a reason for that yet. Either way, we will only deprecate the current wrappers if a new and improved API for low-level access to BLAS/LAPACK functions is available to switch to.

@ilayn I'm all for improving our BLAS/LAPACK wrapping, however it requires some prototyping and a very crisp and concise proposal, after laying out all the constraints that the new situation should meet. This RFC as written should probably be closed, because it is fairly incomplete and as written it cannot be executed, due to the points I raised above.

Maybe let's use this one as a discussion thread, and at some point open a new one which is complete?

**Comment by ilayn** [comment]
> If you deprecate the F2PY wrappers,

If we are going to deprecate wrappers we are not going to remove them without alternatives. The problem with the existing ones are just historical inconsistencies and by quite a lot. There is hardly any consistency of the wrappers and everyone of them feels unique. Some exports m, n array size some hide it and so on many weird stuff as I mentioned above. So we will deprecate when we have a proper alternative which what this issue is about.

> Also ?axpy, which is much faster than single-threaded addition of ndarrays. 

That's very unlikely to beat numpy if same optimizations are given at build time to both.

**Comment by ilayn** [comment]
> This RFC as written should probably be closed, because it is fairly incomplete and as written it cannot be executed, due to the points I raised above.

Not sure what you mean by this. This is an RfC not a PR. And as I am trying to make a case, there are ways to do this including Cython. We just did not choose which way. But I don't agree that backwards compatibility is an issue here because it will be needed at some point because of `const` addition saga and many other things are waiting for similar handling.

**Comment by rgommers** [comment]
An RFC is usually a detailed proposal about what to do. This is more a "hey let's get rid of the `f2py` wrappers" without (so far) a concrete way of achieving that, nor laying out all the constraints/requirements. So I just really don't know what to do with this at the moment beyond the input I have already given in my first comment - it's too vague. All I can say is "sure, I'd love better/simpler wrappers and less duplication", but that's about it. To me, that is not a very useful open issue.

Let's turn this around: what do you actually want as input from other maintainers here? Or what are the next steps?

**Comment by chillenb** [comment]
Thanks everybody for your clarifications!! I'm glad to hear that BLAS wrappers are going to be kept in the public API, at least in some form.

That said, I'd like to push back against the following statement a little:

> > Also ?axpy, which is much faster than single-threaded addition of ndarrays.
> 
> That's very unlikely to beat numpy if same optimizations are given at build time to both.

The multithreading does make a difference! It's hard for a single thread to saturate the memory bandwidth of the entire CPU. This gets worse if there is NUMA. Here is a benchmark on a machine with 2x Intel Xeon 6252N @2.6 GHz. NumPy was compiled with -march=native -mtune=native and BLAS is OpenBLAS from conda-forge:
```
In [2]: X=np.random.random(1000**3)

In [3]: Y=np.random.random(1000**3)

In [4]: %timeit Z=X+Y
3.15 s ± 3.02 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: from scipy.linalg.blas import daxpy

In [7]: %timeit daxpy(X,Y)
656 ms ± 27.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [8]: %timeit np.add(X,Y,out=Y)
1.25 s ± 358 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

With Intel Distribution for Python, even np.add is multithreaded. It's a pretty significant boost:

```
In [5]: %timeit Z=X+Y
339 ms ± 8.14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit daxpy(X,Y)
413 ms ± 53.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: %timeit np.add(X,Y,out=Y)
209 ms ± 49.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by ilayn** [comment]
@chillenb That is because the default OpenBLAS is single-threaded. So it is true that it would make a difference indeed. What I meant to say is that if you also use a BLAS/LAPACK with threading enabled, Numpy does the same thing. Intel distro has multithreaded MKL hence the boost. 

Could you please try also with infix operator `Y += X`?

**Comment by ilayn** [comment]
> Let's turn this around: what do you actually want as input from other maintainers here? Or what are the next steps?

Would that be OK if I remove RfC and put `query` instead? Because I want to discuss the very next steps that are scattered around different issues. I don't think this is an impossible problem at all but needs coordination and will break things. However, we have been delaying these things for a very long time now and leaving a lot of performance at the table through the f2py glue. 

Spending a lot of time on a possible solution and then realizing that we can't do it serves noone and wastes limited time. Your comments are quite valuable in that regard but I don't think they pose an immovable block so far.

**Comment by chillenb** [comment]
> @chillenb That is because the default OpenBLAS is single-threaded. So it is true that it would make a difference indeed. What I meant to say is that if you also use a BLAS/LAPACK with threading enabled, Numpy does the same thing. Intel distro has multithreaded MKL hence the boost. At some point I think NumPy uses daxpy anyways.
> 
> Could you please try also with infix operator `Y += X`?

Sure thing. Here is what I found:


```
# incY() does Y += X

In [14]: %timeit incY()
1.13 s ± 1.57 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

To get more data I also tried conda-forge NumPy, with libblas=MKL, and found this:
```
In [10]: %timeit scipy.linalg.blas.daxpy(X,Y)
539 ms ± 15.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [11]: %timeit np.add(X,Y,out=Y)
1.22 s ± 1.43 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [19]: %timeit incY()
1.29 s ± 150 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Both OpenBLAS and MKL ate 100% of the CPU, but np.add(X,Y,out=Y) and X+=Y were both single-threaded. This suggests np.add did not touch BLAS at all.

To be sure of this, I also checked MKL's verbose mode:
```
>>> X=np.random.random(1024**2)
>>> Y=np.random.random(1024**2)
>>> np.add(X,Y,out=Y)
array([1.34086246, 0.20451872, 0.92893604, ..., 0.51186964, 0.59766295,
       0.39522988])
>>> scipy.linalg.blas.daxpy(X,Y)
MKL_VERBOSE DAXPY(1048576,0x7ffc0562e8b8,0x7dfe5d200010,1,0x7dfe5c800010,1) 1.26ms CNR:OFF Dyn:1 FastMM:1 TID:0  NThr:48
array([1.97819284, 0.20506346, 1.21871963, ..., 0.72971799, 0.8775876 ,
       0.61443089])
>>> 
```

**Comment by ilayn** [comment]
> I think this alone is a blocker until the relevant Fortran code is gone (xref https://github.com/scipy/scipy/issues/18566).

Yes I'm working on it, hang in there 😅 I know it is really doable  in one sit for many libs but work is crazy these days. Already a bit bummed out missing 1.14

But lapack wrappers are party of #18566 just saying. This is hence pre work for that

**Comment by ilayn** [comment]
I basically re-read this again and I think the discussion derailed at some point. I understand Ralf's concern but I don't see the actual technical point since we did this with multiple libraries already. The query here (which I still think is RfC) is using the already existing architecture to provide Python level functions. 

I don't see how symbol names etc. came into picture here. Basically, we have low-level access to BLAS/LAPACK functions. Instead of going through `f2py` we would go with Cython/C route which we have been already doing. Maybe I don't understand the objection here.

**Comment by lucascolley** [comment]
> The only real problem is to whether we should break the wrappers because they are in quite bad shape in terms of inconsistencies so whether this is a SciPy 2.0 thing or not should be established

Should we use this issue to discuss just internal changes first? Any 'SciPy 2.0' things seem to depend on public API changes, right?

**Comment by ilayn** [comment]
ILP64 issue is how we switch to 64-bit integers in BLAS/LAPACK calls. This issue is about how we provide Python API for making to users (IPL64 or not). They are a bit orthogonal.

**Comment by rgommers** [comment]
The problem is that there are too many things mixed in this issue, and your initial proposal is unclear. We have to separate a number of things here:

1. Exposing ILP32 bindings as _public_ API for other Cython-using packages: this is `cython_blas`, `cython_lapack`
2. Exposing raw BLAS/LAPACK functions as _public_ Python API: this is `linalg.blas`, `linalg.lapack`
3. How to deal with BLAS/LAPACK bindings _internally_ from C, C++ code: a bit too custom now, covered in gh-21130
4. How to deal with BLAS/LAPACK bindings _internally_ from Cython code: this one is settled, that's `cython_blas`, `cython_lapack`
5. How to deal with BLAS/LAPACK bindings _internally_ from pure Python code: currently this is via `linalg.blas`, `linalg.lapack`
6. What to do about ILP64 BLAS/LAPACK: this is gh-21889, so let's keep that out of this issue
7. The Fortran removal effort: not relevant to this issue at all, no Fortran code nor Fortran compiler is involved
8. Stop relying on `numpy.f2py`: maybe the key driver for this issue? If so, can you please explicitly say that? Can be done, but then let's be clear how it impacts items 1-5.

**Comment by ev-br** [comment]
> 6. What to do about ILP64 BLAS/LAPACK: this is RFC: 64-bit LAPACK and cython_lapack #21889, so let's keep that out of this issue

Note that in the ILP64-aware world, `linalg.lapack` is currently the only way to transparently switch between LP64 and ILP64 variants at runtime, via `get_lapack_funcs(..., use_ilp64="preferred")`. All other ways are build-time only.

> 8. Stop relying on numpy.f2py: maybe the key driver for this issue? If so, can you please explicitly say that? Can be done, but then let's be clear how it impacts items 1-5.

Cross-ref https://github.com/scipy/scipy/pull/22924 et al. Once we converge on implementation details (hint hint Ilhan :-))), there's a clear path towards high-level functions, `inv`, `solve` etc, to not rely on `numpy.f2py` anymore.

Re: impacts on 1-5: gh-22924 way is orthogonal to all of these.

**Comment by ilayn** [comment]
I am a bit confused actually about your response. This issue only and exclusively mentions 8. I don't quite see where the others come in. Because I never mentioned any of those involved. Maybe you have a more elaborate issue in mind? I think we need to be a bit more careful before we start typing past each other. 

This is strictly about exposing Python API from `scipy.linalg.lapack` and `scipy.linalg.blas` only changing their underlying binding from Fortran generated functions to Cython or directly C generated . So in your enumeration, though not exactly mapping, item 5 using item 6. 

The ILP64 situation is about how to expose the ILP64 enabled versions of the same functions. But it is opaque here since the Python wrappers don't have an `int` restriction. That's why it is orthogonal in my opinion but here is an example. 

Currently we have `scipy.linalg.blas.dnrm2` function wrapped `n2 = dnrm2(x,[n,offx,incx])`. This one through a fortran run and via our fblas.pyf.src files exposes the BLAS function. However it is already exposed through `scipy.linalg.cython_blas` interface. So we are exposing it twice and moreover putting fortran compilation in between, both complicating the build and increasing the build size. 

We can instead have a function that exposes to Python API a blas function 


```Cython
def dnrm2(.... ):
    ...some declarations

    < the crucial part that f2py convenience provides>
        Make sure that all is Fortran contiguous
        Make sure all scalars are properly casted

```

This will not touch the way we expose `scipy.linalg.blas.dnrm2` but will change where it calls the underlying BLAS call. 

Is this a bit clearer?

**Comment by ilayn** [comment]
Not to jampack more things into the example above, the SciPy 2.0 issue is the following; 

Historically, we had a lot of wrappers many involving the wrapper author's opinions seeping into the code. For example as you can see above, original `dnrm2` does not have an offset but our wrapper has for some reason. 

More so in premutation array `ipiv` manipulations, hoping to make the life of the Python user's life easier to convert to 0-indexes. But it is not done comprehensively so sometimes there sometimes not. 

https://github.com/scipy/scipy/blob/4674ab33af519eddc34709c9f75106f95eaf0482/scipy/linalg/flapack_gen.pyf.src#L89

Sometimes `uplo` is a char sometimes it is mapped to a keyword argument `lower=<1, 0>`. This one is actually all over the place with no visible logic. And so on. 

So, **if we make the switch**,  We can fix all these which will make the auto-generation much easier. Otherwise we have to hand-wrap existing functions just like we have Fortran wrappers. It's not impossible just tedious.

Hence the SciPy 2.0 mention is only for thinking about fixing these inconsistencies or pushing to future and switching to Cython based wrappers with the historicals warts carried over.

**Comment by rgommers** [comment]
> This will not touch the way we expose `scipy.linalg.blas.dnrm2` but will change where it calls the underlying BLAS call.
>
> Is this a bit clearer?

Yes, thanks, that is clear. This wouldn't be so hard to write for a few functions by hand, right? Say pick two functions: one BLAS function that you want to remain unchanged, and one you'd perhaps want to change the signature of . Or perhaps you give up on changing any signatures at all, and then the story is even cleaner. We can then just look at the new code, and if it'd be a drop-in replacement for `linalg.blas`, and what it buys us.

> So we are exposing it twice and moreover putting fortran compilation in between

Yes, we are building two sets of wrappers, which is slightly wasteful (binary size wise; compile time isn't much). But no we aren't putting Fortran in between, there is zero Fortran code going into `_fblas.so`/`_flapack.so`. The extension name may be a bit misleading - it's pure C (unless I'm being really dense, but if so please point out the exact line of code in a build file where we hit Fortran code).

> So, **if we make the switch**, We can fix all these which will make the auto-generation much easier. Otherwise we have to hand-wrap existing functions just like we have Fortran wrappers. It's not impossible just tedious.

It's a lot of functions to think about. I'm not sure we can do it before having the code. Why not go with the easy-to-write version first, put it right next to the old version, and then we can assess? Mismatches will affect a limited number of users, and they're the type of users that can help themselves.

**Comment by ilayn** [comment]
This is not required as per the discussion in #24014.
